[Ready for Review/Testmerge]Adds scanner gates

XDTM opened the pull request:
:cl: XDTM
add: Added scanner gates: these buildable machines will scan any mob passing under it, and it will alert when a set condition is met.
add: Possible conditions: Wanted status, Mindshield, Nanites, Diseases, Species, or carrying guns.
add: Can be inverted, so it will alert when the condition is _not_ met, and can be id-locked to avoid tampering.
/:cl:

Currently ready and tested, but there's room for improvement (signaller integration, directional icons, etc.)

XDTM committed:
Adds scanner gates

81Denton commented:
I have some ideas:
- Allow attaching signaler assemblies, for example for pulsing airlock wires, bombs, circuitry
- AND/OR/XOR etc. logic gates, for example "alert if lizard with gun but no mindshield passes through this"
- Radiation scanning
- Scanning the ID's assigned job
- Does get_disease_severity_value respect the virus' stealth level?

ExcessiveUseOfCobblestone commented on code/datums/diseases/_disease.dm:
```diff
@@ -160,3 +160,21 @@
+		if(DISEASE_SEVERITY_DANGEROUS)
+			return 6
+		if(DISEASE_SEVERITY_BIOHAZARD)
+			return 7
```
isn't this the purpose of defines...

XDTM replied:
yes but those are text defines

MrDoomBringer replied:
just make a dictionary smhfam

ninjanomnom replied:
@MrDoomBringer is correct here

XDTM committed:
fix

MrDoomBringer commented:
NICE
i was just playing goon the other day and wishing we had these

ExcessiveUseOfCobblestone commented:
@81Denton for your last question: yes per https://github.com/tgstation/tgstation/pull/39986/files#diff-98e1663438f2152652cd609c8987315fR97

It's the same proc used for medhuds, PENLITEs, and now scanners

JJRcop commented on code/game/machinery/scan_gate.dm:
```diff
@@ -0,0 +1,195 @@
+						beep = TRUE
+		if(SCANGATE_GUNS))
+			for(var/I in M.get_contents())
+				if(istype(I, /obj/item/gun))
```
This will go off with wands and staves, blacklist /obj/item/gun/magic?

XDTM replied:
I thought it would be a lot of extra processing to do so, so i kept it simple

JJRcop replied:
You could create a static typecache of /obj/item/gun, then subtract `typesof(/obj/item/gun/magic)` from the typecache.

XDTM replied:
that could work, it might allow me to make other types of contraband scans

XDTM committed:
fix

XDTM committed:
fix2

XDTM committed:
fix3

MimicFaux commented:
Could this also be made to detect romerol/xenomorph infections?

XDTM commented:
zombie species detects romerol infections

JJRcop commented on code/game/machinery/scan_gate.dm:
```diff
@@ -0,0 +1,197 @@
+		if(SCANGATE_WANTED)
+			if(iscarbon(M))
+				var/mob/living/carbon/C = M
+				var/perpname = M.get_face_name(M.get_id_name())
```
Gotta use C

F-OS commented on code/game/machinery/scan_gate.dm:
```diff
@@ -0,0 +1,197 @@
+				disease_threshold = new_threshold
+			. = TRUE
+		//Some species are not scannable, like abductors (too unknown), androids (too artificial) or skeletons (too magic)
+		if("set_target_species")
```
racist

XDTM committed:
update!

XDTM committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into ScanGates

XDTM commented:
- Added icon, made by @Nabski89 
- Added a new sound as the alarm
- Added the option to specify a specific nanite cloud ID to check
- Fixed a couple HUD bugs
- FIxed a small bug where captain's ID was missing an access present in the cargo techs' ID (and thus they could lock the scanner without the captain being able to unlock it)

AnturK commented on code/game/machinery/scan_gate.dm:
```diff
@@ -0,0 +1,226 @@
+	deltimer(scanline_timer)
+	add_overlay(type)
+	if(duration)
+		scanline_timer = addtimer(CALLBACK(src, .proc/set_scanline, "passive"), duration, TIMER_STOPPABLE)
```
Is there something special simple flick or state change won't do here ? Cutting all overlays seems excessive.

XDTM replied:
I guess i could store the scanline overlay in a var, i was following the example of other update_icon procs

AnturK replied:
It's not horrible, just could be optimized probably.

AnturK commented on code/game/machinery/scan_gate.dm:
```diff
@@ -0,0 +1,226 @@
+			reverse = !reverse
+			. = TRUE
+		if("set_disease_threshold")
+			var/new_threshold = input("Set disease threshold","Scan Mode") as null|anything in list(DISEASE_SEVERITY_POSITIVE,
```
You don't do any intractability validation after the inputs here.

AnturK commented on code/modules/jobs/access.dm:
```diff
@@ -127,7 +127,7 @@
 	            ACCESS_TELEPORTER, ACCESS_EVA, ACCESS_HEADS, ACCESS_CAPTAIN, ACCESS_ALL_PERSONAL_LOCKERS,
 	            ACCESS_TECH_STORAGE, ACCESS_CHAPEL_OFFICE, ACCESS_ATMOSPHERICS, ACCESS_KITCHEN,
-	            ACCESS_BAR, ACCESS_JANITOR, ACCESS_CREMATORIUM, ACCESS_ROBOTICS, ACCESS_CARGO, ACCESS_CONSTRUCTION,
+	            ACCESS_BAR, ACCESS_JANITOR, ACCESS_CREMATORIUM, ACCESS_ROBOTICS, ACCESS_CARGO, ACCESS_CARGO_BOT, ACCESS_CONSTRUCTION,
```
Was this something missing or some merge issue ?

XDTM replied:
Noticed the bug while testing

AnturK commented on code/modules/mob/living/carbon/human/species_types/jellypeople.dm:
```diff
@@ -1,6 +1,6 @@
 /datum/species/jelly
 	// Entirely alien beings that seem to be made entirely out of gel. They have three eyes and a skeleton visible within them.
-	name = "Xenobiological Jelly Entity"
+	name = "Jellyperson"
```
Seems completely unrelated.

Iamgoofball replied:
name on the scanner stuff prolly

XDTM replied:
Yeah, i use the species.name for the display name, and this didn't matter before since it was invisible ingame

ExcessiveUseOfCobblestone replied:
I actually like the first one better lol

XDTM replied:
It is nice but base jellypeople are no longer a thing, so it's better to have a catch-all name for jelly species in this case.

AnturK reviewed (commented):
Code looks fine, just some questions.

AnturK commented:
Rebuild tgui and this should be good.

AnturK commented:
@XDTM Still alive ?

XDTM committed:
tgui build

XDTM committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into ScanGates

XDTM commented:
sorry, i keep forgetting about reviews
